pkg/controller/async_backup_operations_controller.go

```diff
@@ -0,0 +1,514 @@
+		}
+		err2 := c.updateBackupAndOperationsJSON(ctx, original, backup, nil, &operationsForBackup{errsSinceUpdate: []string{err.Error()}}, false, false)
+		if err2 != nil {
+			log.WithError(err2).Error("error updating Backup")
+		}
+		return ctrl.Result{}, errors.Wrap(err, "error getting backup storage location")
+	}
+
+	if loc.Spec.AccessMode == velerov1api.BackupStorageLocationAccessModeReadOnly {
+		log.Infof("Cannot check progress on async Backup operations because backup storage location %s is currently in read-only mode; marking backup PartiallyFailed", loc.Name)
+		backup.Status.Phase = velerov1api.BackupPhasePartiallyFailed
+
+		err := c.updateBackupAndOperationsJSON(ctx, original, backup, nil, &operationsForBackup{errsSinceUpdate: []string{"BSL is read-only"}}, false, false)
+		if err != nil {
+			log.WithError(err).Error("error updating Backup")
+		}
+		return ctrl.Result{}, nil
+	}
+
+	pluginManager := c.newPluginManager(c.logger)
+	defer pluginManager.CleanupClients()
+	backupStore, err := c.backupStoreGetter.Get(loc, pluginManager, c.logger)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "error getting backup store")
+	}
+
+	operations, err := c.getOperationsForBackup(backupStore, backup.Name)
+	if err != nil {
+		err2 := c.updateBackupAndOperationsJSON(ctx, original, backup, backupStore, &operationsForBackup{errsSinceUpdate: []string{err.Error()}}, false, false)
+		if err2 != nil {
+			return ctrl.Result{}, errors.Wrap(err2, "error updating Backup")
+		}
+		return ctrl.Result{}, errors.Wrap(err, "error getting backup operations")
+	}
+	stillInProgress, changes, opsCompleted, opsFailed, errs := getBackupItemOperationProgress(backup, pluginManager, operations.operations)
+	// if len(errs)>0, need to update backup errors and error log
+	operations.errsSinceUpdate = append(operations.errsSinceUpdate, errs...)
+	backup.Status.Errors += len(operations.errsSinceUpdate)
+	asyncCompletionChanges := false
+	if backup.Status.AsyncBackupItemOperationsCompleted != opsCompleted || backup.Status.AsyncBackupItemOperationsFailed != opsFailed {
+		asyncCompletionChanges = true
+		backup.Status.AsyncBackupItemOperationsCompleted = opsCompleted
+		backup.Status.AsyncBackupItemOperationsFailed = opsFailed
+	}
+	if changes {
+		operations.changesSinceUpdate = true
+	}
+
+	// if stillInProgress is false, backup is complete and needs update
+	// if operations.errsSinceUpdate is not empty, then backup phase needs to change to
+	// BackupPhaseWaitingForPluginOperationsPartiallyFailed and needs update
+	// If the only changes are incremental progress, then no write is necessary, progress can remain in memory
+	if !stillInProgress {
+		if len(operations.errsSinceUpdate) > 0 {
+			backup.Status.Phase = velerov1api.BackupPhaseWaitingForPluginOperationsPartiallyFailed
+		}
+		hasItemsForUpdate := operations.anyItemsToUpdate()
+		if backup.Status.Phase == velerov1api.BackupPhaseWaitingForPluginOperations && hasItemsForUpdate {
```

Comment:
Following the `general finalizing phase/handler` concept as I mentioned in another comment, I suggest we set every backup to `BackupPhaseFinalizingAfterPluginOperations` as long as all their async operations complete, we let the backup to decide whether it has anything to process in the finalizing handler, if nothing is required, it set its own status to `Complete`